pkg/controllers/machine_controller.go

```diff
@@ -312,92 +312,97 @@ func (r *MachineReconciler) startGarbageCollector(ctx context.Context, log logr.
+			if err := r.processMachineDeletion(ctx, logger, machine); err != nil {
 				logger.Error(err, "failed to garbage collect machine")
 			}
 		}
 
 	}, r.resyncIntervalGarbageCollector)
 }
 
-func (r *MachineReconciler) processMachineDeletion(ctx context.Context, log logr.Logger, machine *api.Machine, timeout time.Duration) error {
-	if !machine.Spec.ShutdownAt.IsZero() && time.Now().After(machine.Spec.ShutdownAt.Add(timeout)) {
-		err := r.destroyDomain(log, machine.ID)
-		if err != nil && !libvirt.IsNotFound(err) {
-			return fmt.Errorf("failed to destroy machine domain: %w", err)
-		}
-
-		log.V(1).Info("Destroyed domain")
-		return r.cleanupAdditionalResources(ctx, log, machine)
+func (r *MachineReconciler) processMachineDeletion(ctx context.Context, log logr.Logger, machine *api.Machine) error {
+	isDeleting, err := r.deleteMachine(ctx, log, machine)
+	switch {
+	case isDeleting:
+		return nil
+	case err != nil:
+		return fmt.Errorf("failed to delete machine: %w", err)
 	}
+	log.V(1).Info("Deleted machine")
 
-	err := r.shutdownMachine(ctx, log, machine)
-	if err != nil {
-		if libvirt.IsNotFound(err) {
-			return r.cleanupAdditionalResources(ctx, log, machine)
-		}
-		return fmt.Errorf("failed to shutdown machine: %w", err)
+	if err := r.deleteVolumes(ctx, log, machine); err != nil {
+		return fmt.Errorf("failed to remove machine disks: %w", err)
 	}
+	log.V(1).Info("Removed machine disks")
 
-	return nil
-}
+	if err := r.deleteNetworkInterfaces(ctx, log, machine); err != nil {
+		return fmt.Errorf("failed to remove machine network interfaces: %w", err)
+	}
+	log.V(1).Info("Removed network interfaces")
 
-func (r *MachineReconciler) destroyDomain(log logr.Logger, machineID string) error {
-	log.V(1).Info("Starting domain destroy")
+	if err := os.RemoveAll(r.host.MachineDir(machine.ID)); err != nil {
+		return fmt.Errorf("failed to remove machine directory: %w", err)
+	}
+	log.V(1).Info("Removed machine directory")
 
-	domain := libvirt.Domain{
-		UUID: libvirtutils.UUIDStringToBytes(machineID),
+	machine.Finalizers = utils.DeleteSliceElement(machine.Finalizers, MachineFinalizer)
+	if _, err := r.machines.Update(ctx, machine); store.IgnoreErrNotFound(err) != nil {
+		return fmt.Errorf("failed to update machine metadata: %w", err)
 	}
+	log.V(1).Info("Removed Finalizers")
 
-	return r.libvirt.DomainDestroy(domain)
+	return nil
 }
 
-func (r *MachineReconciler) shutdownMachine(ctx context.Context, log logr.Logger, machine *api.Machine) error {
-	log.V(1).Info("Shutting down Machine")
+func (r *MachineReconciler) deleteMachine(ctx context.Context, log logr.Logger, machine *api.Machine) (bool, error) {
+	domain := libvirt.Domain{
+		UUID: libvirtutils.UUIDStringToBytes(machine.ID),
+	}
 
 	if machine.Spec.ShutdownAt.IsZero() {
-		machine.Spec.ShutdownAt = time.Now()
-		if _, err := r.machines.Update(ctx, machine); err != nil {
-			return fmt.Errorf("failed to set ShutdownAt: %w", err)
-		}
+		return r.shutdownMachine(ctx, log, machine, domain)
```

Comment:
call shutdown multiple times can be useful what i heard from @gehoern because sometimes signal can be ignore.